### ISSUE TITLE
GPU Nodepool Node not Registered

### ISSUE THREAD

**Issue by tospe** [Azure/karpenter-provider-azure#181]
### Version

__Kubernetes Version:__ v1.27.9

### Expected Behavior

When a pod is pending a node claim is successfully created.

### Actual Behavior

GPU Node claim is created however the status is always false. When described we get the message "Node not registered with cluster"

### Steps to Reproduce the Problem

- Create a brand new aks with karpenter support (https://learn.microsoft.com/en-us/azure/aks/node-autoprovision?tabs=azure-arm#install-the-aks-preview-cli-extension)
- created the nodepool gpu-spot and the aksnodeclass
- installed the device plugin daemonset (https://github.com/Azure/karpenter-provider-azure/blob/main/examples/workloads/device-plugin.yaml)
- created sample mnist job (https://github.com/Azure/karpenter-provider-azure/blob/main/examples/workloads/samples-mnist.yaml)

### Resource Specs and Logs

<details>
<summary>Click to expand</summary>

Name:         gpu-spot-57k4r
Namespace:    
Labels:       karpenter.azure.com/sku-accelerator=T4
              karpenter.azure.com/sku-cpu=4
              karpenter.azure.com/sku-encryptionathost-capable=true
              karpenter.azure.com/sku-family=N
              karpenter.azure.com/sku-gpu-count=1
              karpenter.azure.com/sku-gpu-manufacturer=nvidia
              karpenter.azure.com/sku-gpu-name=T4
              karpenter.azure.com/sku-memory=28672
              karpenter.azure.com/sku-name=Standard_NC4as_T4_v3
              karpenter.azure.com/sku-networking-accelerated=true
              karpenter.azure.com/sku-storage-ephemeralos-maxsize=188.978561024
              karpenter.azure.com/sku-storage-premium-capable=true
              karpenter.azure.com/sku-version=3
              karpenter.azure.com/zone=westeurope-2
              karpenter.sh/capacity-type=spot
              karpenter.sh/nodepool=gpu-spot
              kubernetes.io/arch=amd64
              kubernetes.io/os=linux
              node.kubernetes.io/instance-type=Standard_NC4as_T4_v3
              topology.kubernetes.io/region=westeurope
Annotations:  karpenter.azure.com/in-place-update-hash: 243595438
              karpenter.sh/nodepool-hash: 14131116538321825039
API Version:  karpenter.sh/v1beta1
Kind:         NodeClaim
Metadata:
  Creation Timestamp:  2024-03-06T18:48:55Z
  Finalizers:
    karpenter.sh/termination
  Generate Name:  gpu-spot-
  Generation:     1
  Managed Fields:
    API Version:  karpenter.sh/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:karpenter.azure.com/in-place-update-hash:
          f:karpenter.sh/nodepool-hash:
        f:finalizers:
          .:
          v:"karpenter.sh/termination":
        f:generateName:
        f:labels:
          .:
          f:karpenter.azure.com/sku-accelerator:
          f:karpenter.azure.com/sku-cpu:
          f:karpenter.azure.com/sku-encryptionathost-capable:
          f:karpenter.azure.com/sku-family:
          f:karpenter.azure.com/sku-gpu-count:
          f:karpenter.azure.com/sku-gpu-manufacturer:
          f:karpenter.azure.com/sku-gpu-name:
          f:karpenter.azure.com/sku-memory:
          f:karpenter.azure.com/sku-name:
          f:karpenter.azure.com/sku-networking-accelerated:
          f:karpenter.azure.com/sku-storage-ephemeralos-maxsize:
          f:karpenter.azure.com/sku-storage-premium-capable:
          f:karpenter.azure.com/sku-version:
          f:karpenter.azure.com/zone:
          f:karpenter.sh/capacity-type:
          f:karpenter.sh/nodepool:
          f:kubernetes.io/arch:
          f:kubernetes.io/os:
          f:node.kubernetes.io/instance-type:
          f:topology.kubernetes.io/region:
        f:ownerReferences:
          .:
          k:{"uid":"486ae212-581e-4ae5-b468-cdc8a8d60f89"}:
      f:spec:
        .:
        f:nodeClassRef:
          .:
          f:name:
        f:requirements:
        f:resources:
          .:
          f:requests:
            .:
            f:cpu:
            f:memory:
            f:nvidia.com/gpu:
            f:pods:
        f:taints:
    Manager:      karpenter
    Operation:    Update
    Time:         2024-03-06T18:50:17Z
    API Version:  karpenter.sh/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:allocatable:
          .:
          f:cpu:
          f:ephemeral-storage:
          f:memory:
          f:nvidia.com/gpu:
          f:pods:
        f:capacity:
          .:
          f:cpu:
          f:ephemeral-storage:
          f:memory:
          f:nvidia.com/gpu:
          f:pods:
        f:conditions:
        f:providerID:
    Manager:      karpenter
    Operation:    Update
    Subresource:  status
    Time:         2024-03-06T18:50:17Z
  Owner References:
    API Version:           karpenter.sh/v1beta1
    Block Owner Deletion:  true
    Kind:                  NodePool
    Name:                  gpu-spot
    UID:                   486ae212-581e-4ae5-b468-cdc8a8d60f89
  Resource Version:        93625
  UID:                     5e541c74-238a-4ba5-a302-dc80ca15e89a
Spec:
  Node Class Ref:
    Name:  gpu-spot
  Requirements:
    Key:       karpenter.sh/nodepool
    Operator:  In
    Values:
      gpu-spot
    Key:       node.kubernetes.io/instance-type
    Operator:  In
    Values:
      Standard_NC4as_T4_v3
    Key:       kubernetes.io/arch
    Operator:  In
    Values:
      amd64
    Key:       kubernetes.io/os
    Operator:  In
    Values:
      linux
    Key:       karpenter.azure.com/sku-name
    Operator:  In
    Values:
      Standard_NC4as_T4_v3
  Resources:
    Requests:
      Cpu:             250m
      Memory:          470Mi
      nvidia.com/gpu:  1
      Pods:            7
  Taints:
    Effect:  NoSchedule
    Key:     nvidia.com/gpu
    Value:   true
Status:
  Allocatable:
    Cpu:                  3860m
    Ephemeral - Storage:  128G
    Memory:               22372Mi
    nvidia.com/gpu:       1
    Pods:                 110
  Capacity:
    Cpu:                  4
    Ephemeral - Storage:  128G
    Memory:               26521Mi
    nvidia.com/gpu:       1
    Pods:                 110
  Conditions:
    Last Transition Time:  2024-03-06T18:50:17Z
    Message:               Node not registered with cluster
    Reason:                NodeNotFound
    Status:                False
    Type:                  Initialized
    Last Transition Time:  2024-03-06T18:50:17Z
    Status:                True
    Type:                  Launched
    Last Transition Time:  2024-03-06T18:50:17Z
    Message:               Node not registered with cluster
    Reason:                NodeNotFound
    Status:                False
    Type:                  Ready
    Last Transition Time:  2024-03-06T18:50:17Z
    Message:               Node not registered with cluster
    Reason:                NodeNotFound
    Status:                False
    Type:                  Registered

---

apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: gpu-spot
  annotations:
    kubernetes.io/description: "General purpose NodePool for generic workloads"
spec:
  disruption:
    expireAfter: Never
  template:
    spec:
      requirements:
        - key: kubernetes.io/arch
          operator: In
          values: ["amd64"]
        - key: kubernetes.io/os
          operator: In
          values: ["linux"]
        - key: karpenter.azure.com/sku-name
          operator: In
          values: ["Standard_NC4as_T4_v3"]
      nodeClassRef:
        name: gpu-spot
      taints:
        - key: nvidia.com/gpu
          value: "true"
          effect: NoSchedule

---
apiVersion: karpenter.azure.com/v1alpha2
kind: AKSNodeClass
metadata:
  name: gpu-spot
  annotations:
    kubernetes.io/description: "General purpose AKSNodeClass for running Ubuntu2204 nodes"
spec:
  imageFamily: Ubuntu2204
</details>

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by Bryce-Soghigian** [comment]
Hi @tospe, we are rolling out a release right now that should include a fix for this. 

This is the PR: https://github.com/Azure/karpenter-provider-azure/pull/76 


As a quick mitigation, you might try modifying the AKSNodeClass and use an older image version
[Node autoprovisioning (Preview) - Azure Kubernetes Service | Microsoft Learn](https://learn.microsoft.com/en-gb/azure/aks/node-autoprovision?tabs=azure-cli#node-pools)
spec:
  imageFamily: Ubuntu2204
  imageVersion: 202311.07.0 <--- This Node image should do the trick
  osDiskSizeGB: 128

**Comment by Bryce-Soghigian** [comment]
TLDR, the GPU Image tries to pull the gpu driver image, but the node image required a new sha field to be specified. Please let me know if the mitigation works! Otherwise you can wait for the release to rollout to your region

**Comment by tospe** [comment]
Since a few days ago this issue appeared again. 

Steps to reproduce:

- create a new aks: `az aks create --name karpenter2 --resource-group TECH-SW-AIML-PROD  --node-provisioning-mode Auto --network-plugin azure --network-plugin-mode overlay --network-dataplane cilium`
- apply device plugin: `kubectl apply -f https://raw.githubusercontent.com/Azure/karpenter-provider-azure/main/examples/workloads/device-plugin.yaml`
- create nodepools: see manifests below
- create gpu job: `kubectl apply -f https://raw.githubusercontent.com/Azure/karpenter-provider-azure/main/examples/workloads/samples-mnist.yaml`

It seems that the nvidia drivers are not being installed.

What be 
nodepool manifests 
```
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: gpu-spot
  annotations:
    kubernetes.io/description: "General GPU purpose NodePool for ML workloads"
spec:
  disruption:
    expireAfter: 1h
  template:
    spec:
      requirements:
        - key: kubernetes.io/arch
          operator: In
          values: ["amd64"]
        - key: kubernetes.io/os
          operator: In
          values: ["linux"]
        - key: karpenter.sh/capacity-type
          operator: In
          values: ["spot"]
        - key: karpenter.azure.com/sku-gpu-manufacturer
          operator: In
          values: ["nvidia"]
        - key: topology.kubernetes.io/zone
          operator: In
          values: ["westeurope-1", "westeurope-2", "westeurope-3"]
      nodeClassRef:
        name: gpu-spot
      taints:
        - key: nvidia.com/gpu
          value: "true"
          effect: NoSchedule

---
apiVersion: karpenter.azure.com/v1alpha2
kind: AKSNodeClass
metadata:
  name: gpu-spot
  annotations:
    kubernetes.io/description: "General purpose AKSNodeClass for running Ubuntu2204 nodes"
spec:
  imageFamily: Ubuntu2204
  imageVersion: 202311.07.0
  osDiskSizeGB: 128
```

**Comment by Bryce-Soghigian** [comment]
@tospe  is it possible its just spot being spot and evicting quickly? Can you reproduce this problem with on-demand? Spot capacity for GPUs right now is at an all time low due to AI goldrush. 

Let me try with your configuration as well. Note that the GPU bug fix should have rolled out to westeurope so you shouldn't need the old image version anymore if you are deploying there. 

Regardless thank you for investigating!!! Will look into it more

**Comment by tallaxes** [comment]
> ```yaml
>        - key: topology.kubernetes.io/zone
>          operator: In
>          values: ["westeurope-1", "westeurope-2", "westeurope-3"]
> ```

@tospe, also note that zone constraint currently causes continuous drift: https://github.com/Azure/karpenter-provider-azure/issues/205. (You should not need to specify it, though.)

**Comment by tospe** [comment]
> @tospe is it possible its just spot being spot and evicting quickly? Can you reproduce this problem with on-demand? Spot capacity for GPUs right now is at an all time low due to AI goldrush.
> 
> Let me try with your configuration as well. Note that the GPU bug fix should have rolled out to westeurope so you shouldn't need the old image version anymore if you are deploying there.
> 
> Regardless thank you for investigating!!! Will look into it more

Maybe. I will look into this.

I will close this. 

Currently I am using this to run a ML workloads. If something new appears i will reopen.

**Comment by tospe** [comment]
I am facing this issue again. 

It works in the first hours (<8) and then just stops working. I am not even using spot instances anymore,

Any idea why this might be happening? 

I am using: nvcr.io/nvidia/k8s-device-plugin:v0.14.1
And the following configuration:

```
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: gpu-spot
  annotations:
    kubernetes.io/description: "General GPU purpose NodePool for ML workloads"
spec:
  disruption:
    expireAfter: 1h
  template:
    spec:
      requirements:
        - key: kubernetes.io/arch
          operator: In
          values: ["amd64"]
        - key: kubernetes.io/os
          operator: In
          values: ["linux"]
        - key: karpenter.azure.com/sku-gpu-manufacturer
          operator: In
          values: ["nvidia"]
      nodeClassRef:
        name: gpu-spot
      taints:
        - key: nvidia.com/gpu
          value: "true"
          effect: NoSchedule

---
apiVersion: karpenter.azure.com/v1alpha2
kind: AKSNodeClass
metadata:
  name: gpu-spot
  annotations:
    kubernetes.io/description: "General purpose AKSNodeClass for running Ubuntu2204 nodes"
spec:
  imageFamily: Ubuntu2204
  imageVersion: 202402.12.0

```

**Comment by Bryce-Soghigian** [comment]
@tospe, you should see the nodes as succesfully registering again. 
gpu-spot-7p9pt joined the cluster at 2024-04-03T01:18:27.000Z utc

Note that the nodepool still will select spot vms with that nodepool setup.
"message": "disrupting via consolidation delete, terminating 1 candidates aks-gpu-spot-7p9pt/Standard_NC4as_T4_v3/spot",

We can see here it selected  a spot vm. Its because your nodepool doesn't constrain for on-demand. 
you need 
```
        - key: karpenter.sh/capacity-type
          operator: In
          values: ["on-demand"]
```


Also it would be good to not pin this node image version as we are still in preview so we are experimenting with the community image gallery and may end up deleting some node images. 

Update your yamls to 
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: gpu-spot
  annotations:
    kubernetes.io/description: "General GPU purpose NodePool for ML workloads"
spec:
  disruption:
    expireAfter: 1h
  template:
    spec:
      requirements:
        - key: kubernetes.io/arch
          operator: In
          values: ["amd64"]
        - key: kubernetes.io/os
          operator: In
          values: ["linux"]
        - key: karpenter.azure.com/sku-gpu-manufacturer
          operator: In
          values: ["nvidia"]
        - key: karpenter.sh/capacity-type
          operator: In
          values: ["on-demand"]
      nodeClassRef:
        name: gpu-spot
      taints:
        - key: nvidia.com/gpu
          value: "true"
          effect: NoSchedule

---
apiVersion: karpenter.azure.com/v1alpha2
kind: AKSNodeClass
metadata:
  name: gpu-spot
  annotations:
    kubernetes.io/description: "General purpose AKSNodeClass for running Ubuntu2204 nodes"
spec:
  imageFamily: Ubuntu2204
```. 



RCA as to why this is occurring: NAP has a bug with AKS node image upgrade that causes the nodes to join as unauthorized. A simple put managed cluster(az aks update -g rg -n clusterName) after this upgrade will reconcile the cluster into a happy state. We have the bug fixed for this and it will be rolling out in the current release. If you start to encounter the issue again, that should mitigate it for you. I triggered one myself to mitigate your cluster. 

After the next release rolls out you shouldn't have to worry about this and things will work normally. 

Please apply that nodepool change and let me know if you encounter any other problems!!!!

**Comment by Bryce-Soghigian** [comment]
note that node image upgrade will continue to break the cluster here but mitigation is explained https://github.com/Azure/karpenter-provider-azure/issues/247#issuecomment-2034911505

**Comment by tospe** [comment]
It works! 
If something happens I will try to do the mitigation.

**Comment by HakjunMIN** [comment]
@Bryce-Soghigian
Thank you. It works by using your yaml. Can I specify a sku-family instead? 
like this.
```yaml
requirements:       
  - key: karpenter.azure.com/sku-family
    operator: In
    values: ["NVadsA10v5", "NV"...]
```

**Comment by Bryce-Soghigian** [comment]
@HakjunMIN see the docs on requirements here: https://learn.microsoft.com/en-gb/azure/aks/node-autoprovision?tabs=azure-cli#supported-node-provisioner-requirements

**Comment by HakjunMIN** [comment]
@Bryce-Soghigian  Thank you so much!!
